plugins/inputs/aliyuncms/aliyuncms.go

```diff
@@ -0,0 +1,597 @@
+	//Start periodic discovery process
+	s.dt.Start()
+
+	//Special setting for acs_oss project since the API differs
+	if s.Project == "acs_oss" {
+		s.dimensionKey = "BucketName"
+	}
+
+	return nil
+}
+
+// Start is a noop which is required for a *DockerLogs to implement
+// the telegraf.ServiceInput interface
+func (s *AliyunCMS) Start(telegraf.Accumulator) error {
+	return nil
+}
+
+// Gather implements telegraf.Inputs interface
+func (s *AliyunCMS) Gather(acc telegraf.Accumulator) error {
+
+	s.updateWindow(time.Now())
+
+	// limit concurrency or we can easily exhaust user connection limit
+	lmtr := limiter.NewRateLimiter(s.RateLimit, time.Second)
+	defer lmtr.Stop()
+
+	var wg sync.WaitGroup
+	for _, metric := range s.Metrics {
+		//Prepare internal structure with data from discovery
+		s.prepareTagsAndDimensions(metric)
+		wg.Add(len(metric.MetricNames))
+		for _, metricName := range metric.MetricNames {
+
+			<-lmtr.C
+			go func(metricName string, metric *Metric) {
+				defer wg.Done()
+				acc.AddError(s.gatherMetric(acc, metricName, metric))
+			}(metricName, metric)
+		}
+		wg.Wait()
+	}
+
+	return nil
+}
+
+func (s *AliyunCMS) Stop() {
+	if s.dt != nil {
+		s.dt.Stop()
+	}
+}
+
+func (s *AliyunCMS) updateWindow(relativeTo time.Time) {
+
+	//https://help.aliyun.com/document_detail/51936.html?spm=a2c4g.11186623.6.701.54025679zh6wiR
+	//The start and end times are executed in the mode of
+	//opening left and closing right, and startTime cannot be equal
+	//to or greater than endTime.
+
+	windowEnd := relativeTo.Add(-s.Delay.Duration)
+
+	if s.windowEnd.IsZero() {
+		// this is the first run, no window info, so just get a single period
+		s.windowStart = windowEnd.Add(-s.Period.Duration)
+	} else {
+		// subsequent window, start where last window left off
+		s.windowStart = s.windowEnd
+	}
+
+	s.windowEnd = windowEnd
+}
+
+// Gather given metric and emit error
+func (s *AliyunCMS) gatherMetric(acc telegraf.Accumulator, metricName string, metric *Metric) error {
+
+	req := cms.CreateDescribeMetricListRequest()
+	req.Period = strconv.FormatInt(int64(s.Period.Duration.Seconds()), 10)
+	req.MetricName = metricName
+	req.Length = "10000"
+	req.Namespace = s.Project
+	req.EndTime = strconv.FormatInt(s.windowEnd.Unix()*1000, 10)
+	req.StartTime = strconv.FormatInt(s.windowStart.Unix()*1000, 10)
+	req.Dimensions = metric.requestDimensionsStr
+
+	for more := true; more; {
+		resp, err := s.client.DescribeMetricList(req)
+		if err != nil {
+			return errors.Errorf("failed to query metricName list: %v", err)
+		} else if resp.Code != "200" {
+			return errors.Errorf("failed to query metricName list: %v", resp.Message)
+		}
+
+		var datapoints []map[string]interface{}
+		if err = json.Unmarshal([]byte(resp.Datapoints), &datapoints); err != nil {
+			return errors.Errorf("failed to decode response datapoints: %v", err)
+		}
+
+		if len(datapoints) == 0 {
+			lg.logW("No metrics returned from CMS, response msg: %s", resp.Message)
```

Comment:
Probably a good debug level. it should be pretty quiet in normal operation.

Reply:
Moved to debug level